File: BalancedBinaryTree/non-recursive.java
public static boolean isBalancedDFS(TreeNode root){
    HashMap<TreeNode, Integer> visited = new HashMap<TreeNode, Integer>();
    LinkedList<TreeNode> stack = new LinkedList<TreeNode>();
    int path = 0;
    if( root != null ) stack.push(root);
    while( !stack.isEmpty() ){
        TreeNode top = stack.peek();
        if( !visited.containsKey(top) ){
            visited.put(top, ++path);
            if( top.right != null && !visited.containsKey(top.right) ) stack.push(top.right);
            if( top.left  != null && !visited.containsKey(top.left ) ) stack.push(top.left);
        } else {
            int leftLevel  = top.left  == null? path: visited.get(top.left);
            int rightLevel = top.right == null? path: visited.get(top.right);
            if( Math.abs(leftLevel-rightLevel) > 1 ) return false;
            else visited.put(top, Math.max(leftLevel, rightLevel));
            --path;
            stack.pop();
        }
    }
    return true;
}

File: BalancedBinaryTree/non-recursive.cpp
using namespace std;

#define forLess(i, s, e)    for(int i = (s); i < int(e); ++i)
#define forto(i, s, e)  for(int i = (s); i <= int(e); ++i)

struct Node
{
    Node(TreeNode* r) : root(r), flag(false), lh(-1), rh(-1) {}
    TreeNode* root;
    bool flag;
    int lh, rh;
};

class Solution 
{
public:
    bool isBalanced(TreeNode* root) 
    {
        if(root == NULL)
            return true;
        stack<Node*> s;
        TreeNode *p = root;
        while(p != NULL || !s.empty()){
            if(p == NULL){
                Node *n = s.top();
                if(!n->flag){
                    p = n->root->right;
                    n->flag = true;
                }
                else{
                    if(n->root->left == NULL)
                        n->lh = 0;
                    if(n->root->right == NULL)
                        n->rh = 0;
                    if(abs(n->lh - n->rh) > 1)
                        return false;
                    s.pop();
                    if(!s.empty()){
                        int h = max(n->lh, n->rh) + 1;
                        if(!s.top()->flag)
                            s.top()->lh = h;
                        else
                            s.top()->rh = h;
                    }
                }
            }
            else{
                s.push(new Node(p));
                p = p->left;
            }
        }
        return true;
    }
};
